[expo-file-system] Fix compatibility with Android R

lukmccall opened the pull request:
# Why

Followup of https://github.com/expo/expo/pull/11647.

# How

- Handled SAF files in the `expo-file-system`. 
- Added a way to ask for external storage write permissions.

> **Note**: Those changes were needed to ensure a smooth transition to a scoped storage on Android.

# Test Plan

- NCL ✅
```tsx
import * as React from 'react';
import { View, StyleSheet, Button } from 'react-native';

import * as Constants from 'expo-constants';
import * as MediaLibrary from 'expo-media-library';
import * as FileSystem from 'expo-file-system';

export default function App() {
  return (
    <View style={styles.container}>
      <Button
        title="Get permissions"
        onPress={async () => {
          await MediaLibrary.requestPermissionsAsync();
        }}
      />

      <Button
        title="get info of external directory"
        onPress={async () => {
          const info = await FileSystem.getInfoAsync('file:///storage/emulated/0/9GAG');
          console.log(info);
        }}
      />
      <Button
        title="ask for folder permissions"
        onPress={async () => {
          const wasSuccessful = await FileSystem.askForDirectoryPermissions();
          console.log(wasSuccessful);
        }}
      />
      <Button
        title="write to file in external directory"
        onPress={async () => {
          const dir = await FileSystem.readDirectoryAsync(
            'content://com.android.externalstorage.documents/tree/primary%3A9GAG'
          );
          const fileUri = dir.filter(path => path.includes('.txt'))[0];

          await FileSystem.writeAsStringAsync(fileUri, 'Expo is awesome 🚀🚀🚀');
        }}
      />
      <Button
        title="read from external directory"
        onPress={async () => {
          const dir = await FileSystem.readDirectoryAsync(
            'content://com.android.externalstorage.documents/tree/primary%3A9GAG'
          );
          console.log({ dir });
          const fileUri = dir.filter(path => path.includes('.txt'))[0];
          const data = await FileSystem.readAsStringAsync(fileUri);

          console.log({ data });
        }}
      />
      <Button
        title="remove from external directory"
        onPress={async () => {
          const dir = await FileSystem.readDirectoryAsync(
            'content://com.android.externalstorage.documents/tree/primary%3A9GAG'
          );

          const fileUri = dir.filter(path => path.includes('.txt'))[0];
          const data = await FileSystem.deleteAsync(fileUri);

          console.log({ data });
        }}
      />
      <Button
        title="copy to internal directory"
        onPress={async () => {
          const result = await FileSystem.copyAsync({
            from: 'content://com.android.externalstorage.documents/tree/primary%3A9GAG',
            to: FileSystem.documentDirectory,
          });
          console.log({ result });
        }}
      />
      <Button
        title="move to internal directory"
        onPress={async () => {
          const result = await FileSystem.moveAsync({
            from: 'content://com.android.externalstorage.documents/tree/primary%3A9GAG',
            to: FileSystem.documentDirectory,
          });
          console.log({ result });
        }}
      />
    </View>
  );
}
const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    paddingTop: Constants.statusBarHeight,
    backgroundColor: '#ecf0f1',
    padding: 8,
  },
  paragraph: {
    margin: 24,
    fontSize: 18,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  videoContainer: {
    alignSelf: 'center',
    height: 250,
    width: 250,
    backgroundColor: 'red',
  },
});
```

lukmccall committed:
[expo-file-system] Fix compatibility with Android R

tsapeta commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java:
```diff
@@ -470,6 +511,49 @@ public void copyAsync(Map<String, Object> options, Promise promise) {
+    if (documentFile.isDirectory()) {
+      for (DocumentFile file : documentFile.listFiles()) {
+        if (documentFile.getName() == null) {
+          continue;
+        }
+
+        transformFilesFromSAF(file, new File(outputDir, documentFile.getName()), copy);
+
+      }
+
+      if (!copy) {
+        documentFile.delete();
+      }
+
+      return;
+    }
+
+    if (documentFile.getName() == null) {
+      return;
+    }
```
```suggestion
          continue;
        }
        transformFilesFromSAF(file, new File(outputDir, documentFile.getName()), copy);
      }
      if (!copy) {
        documentFile.delete();
      }
      return;
    }
    if (documentFile.getName() == null) {
      return;
    }
```

tsapeta commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java:
```diff
@@ -850,6 +947,68 @@ public void downloadResumablePauseAsync(final String uuid, final Promise promise
+  public void onNewIntent(Intent intent) {
+  }
+
+  private static boolean isSAFUri(Context context, Uri uri) {
```
Maybe it would be good to add some explanation in the comment or be more clear by `isExternalStorageUri`?

tsapeta commented on packages/expo-file-system/src/FileSystem.ts:
```diff
@@ -313,3 +313,14 @@ export class DownloadResumable {
 }
+
+// Android only
+export async function askForDirectoryPermissions(
```
```suggestion
export async function askForDirectoryPermissionsAsync(
```

tsapeta commented on packages/expo-file-system/src/FileSystem.types.ts:
```diff
@@ -143,6 +143,7 @@ export interface ExponentFileSystemModule {
   readonly getContentUriAsync?: PlatformMethod;
   readonly getFreeDiskStorageAsync?: PlatformMethod;
   readonly getTotalDiskCapacityAsync?: PlatformMethod;
+  readonly askForDirectoryPermissions?: PlatformMethod;
```
```suggestion
  readonly askForDirectoryPermissionsAsync?: PlatformMethod;
```

tsapeta commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java:
```diff
@@ -440,11 +474,18 @@ public void copyAsync(Map<String, Object> options, Promise promise) {
+      } else if (isSAFUri(getContext(), fromUri)) {
+        DocumentFile documentFile = getNearestSAFFile(fromUri);
+        if (!documentFile.exists()) {
+          promise.reject(null);
```
Is rejecting with `null` intentional?

bbarthec commented on packages/expo-file-system/src/FileSystem.ts:
```diff
@@ -313,3 +313,14 @@ export class DownloadResumable {
   }
 }
+
+// Android only
```
```suggestion
/**
 * Android only
 */
```

bbarthec reviewed (approved):
Looks ok to me 👍 As long as this test scenario works I'm ok with that 👍 
Code looks good to me 😊

esamelson commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java:
```diff
@@ -325,27 +342,23 @@ public void writeAsStringAsync(String uriStr, String string, Map<String, Object>
     try {
       Uri uri = Uri.parse(uriStr);
       ensurePermission(uri, Permission.WRITE);
-      if ("file".equals(uri.getScheme())) {
```
I assume you removed this check because we can write to SAF urls with the `content://` scheme. Would it be appropriate to check `if ("file".equals(uri.getScheme()) || isSAFUri(getContext(), uri))`? Would be good to still throw a nice error message if someone tries to write to some other unsupported uri (like `asset://...`).

esamelson replied:
great, thanks for the clarification.

esamelson commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java:
```diff
@@ -325,27 +342,23 @@ public void writeAsStringAsync(String uriStr, String string, Map<String, Object>
+          }
         }
-        out.close();
-        promise.resolve(null);
```
I don't think this should be removed outright, even if it works it's much clearer to explicitly call this in the success case

esamelson commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java:
```diff
@@ -410,6 +436,14 @@ public void moveAsync(Map<String, Object> options, Promise promise) {
+          return;
+        }
+        File output = new File(toUri.getPath());
+        transformFilesFromSAF(documentFile, output, false);
```
missing promise.resolve call?

esamelson commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java:
```diff
@@ -440,11 +474,18 @@ public void copyAsync(Map<String, Object> options, Promise promise) {
+          return;
+        }
+        File output = new File(toUri.getPath());
+        transformFilesFromSAF(documentFile, output, true);
```
same, missing promise.resolve

esamelson commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java:
```diff
@@ -470,6 +511,49 @@ public void copyAsync(Map<String, Object> options, Promise promise) {
+          continue;
+        }
+
+        transformFilesFromSAF(file, new File(outputDir, documentFile.getName()), copy);
```
is it intentional that if `documentFile` is a directory, all the contents will be copied directly into `outputDir`, rather than to a directory inside of `outputDir`? it seems like if there are nested directories in `documentFile` this will flatten them.

lukmccall replied:
It will create a new folder in the outputDir for each directory in the input path (see a new file from the outputDir and document name). So the hierarchy should be preserved.

esamelson commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java:
```diff
@@ -513,6 +597,19 @@ public void readDirectoryAsync(String uriStr, Map<String, Object> options, Promi
+        DocumentFile[] children = file.listFiles();
+        List<String> result = new ArrayList<>();
+        for (DocumentFile child : children) {
+          result.add(child.getUri().toString());
```
```suggestion
          result.add(child.getName());
```
looking at the rest of the implementation, it looks like this method should return an array of file names rather than URIs.

lukmccall replied:
That is the desired behavior here. You can't just add a file name to the SAF URI to get a valid SAF URI. This URI type is more complicated. I think we don't want to bother about joining base URI with the file name. That's why I've decided to return a list of uris.

esamelson reviewed (changes requested):
This is looking good, thanks for figuring this all out. A few additional notes:
- please add docs for the new `askForDirectoryPermissionsAsync` method
- also, is this a permission we need to scope in the Expo Go client?
- is it possible to add the tests from the test plan (or something similar) to test-suite, or if not, native-component-list?

lukmccall committed:
Apply requested changes

lukmccall commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java:
```diff
@@ -850,6 +947,68 @@ public void downloadResumablePauseAsync(final String uuid, final Promise promise
+    }
+  }
+
+  @Override
+  public void onNewIntent(Intent intent) {
+  }
```
Yep, to implement an interface for UIManager class.

lukmccall committed:
Add comment to `isSAFUri` function

lukmccall commented:
> also, is this a permission we need to scope in the Expo Go client?

We can but I don't think we should. This is the same situation as it was when we decided not to scope permissions for external directories. 

> is it possible to add the tests from the test plan (or something similar) to test-suite, or if not, native-component-list?

I've added a simple action to NCL that checks `askForDirectoryPermissionsAsync` and `readDirectoryAsync`

brentvatne commented on packages/expo-file-system/src/FileSystem.ts:
```diff
@@ -313,3 +313,16 @@ export class DownloadResumable {
+/**
+ * Android only
+ */
+export async function askForDirectoryPermissionsAsync(
```
```suggestion
export async function requestDirectoryPermissionsAsync(
```

we have used `request` elsewhere for permissions, i think it likely makes most sense to do the same here, unless there was an intentional reason to use `askFor` instead

brentvatne commented on docs/pages/versions/unversioned/sdk/filesystem.md:
```diff
@@ -595,24 +595,40 @@ FileSystem.getTotalDiskCapacityAsync().then(totalDiskCapacity => {
+
+#### Arguments
+
+- **initialFileUrl (_string_)** -- **Optional**. The [SAF URI](#saf-uri) of the directory that the file picker should display when it first loads. **Available only on Android R or higher**.
```
how does a user determine the initialFileUrl? where does that come from?

lukmccall replied:
I've added a function that will generate a SAF URI from the album name.

brentvatne commented on docs/pages/versions/unversioned/sdk/filesystem.md:
```diff
@@ -595,24 +595,40 @@ FileSystem.getTotalDiskCapacityAsync().then(totalDiskCapacity => {
+
+#### Returns
+
+Returns a Promise that resolves to a `false` if permissions weren't granted or `string` that represents a [SAF URI](#saf-uri).
```
maybe we should follow something closer to our standard permission response format:

```
{
  status: 'granted' | 'denied';
  granted: boolean;
  directoryUri: string; // or directoryUrl? 👀 
}
```

lukmccall replied:
The return type was changed, but it doesn't include the `status`  key. I don't like the idea behind the strings comparison here.

brentvatne commented:
some thoughts:

### readDirectoryAsync()

the return value for `readDirectoryAsync()` changes significantly if you are reading from a SAF directory or any other directory - notably, the value returned cannot be appended to the directory path that is being read from (because it's the full uri) and, as you pointed out, if we were to return just the file name that would also not work.

given the previous point, i think we should move `readDirectoryAsync()` for SAF directories to a separate function. my proposal is to namespace it under `StorageAccessFramework`, so:

```
import * as FileSystem from 'expo-file-system';
FileSystem.StorageAccessFramework.readDirectoryAsync(path);
```

alternatively users can write:

```
import { StorageAccessFramework } from 'expo-file-system';
StorageAccessFramework.readDirectoryAsync(path);
```

if you call `readDirectoryAsync` on a SAF path we should reject with a useful error

```
import * as FileSystem from 'expo-file-system';
await FileSystem.StorageAccessFramework.readDirectoryAsync(safPath); // cool, good
await FileSystem.readDirectoryAsync(safPath); // error, can't read Storage Access Framework directory, use StorageAccessFramework.readDirectoryAsync() instead
```

### requestDirectoryPermisssionsAsync()

i made a couple suggestions above to change the name to the one used in this heading and also to update the return value. a third suggestion is to move it under the `StorageAccessFramework` namespacing within expo-file-system

```
import { StorageAccessFramework } from 'expo-file-system';
StorageAccessFramework.requestDirectoryPermisssionsAsync(path);
```

### about namespacing to StorageAccessFramework

we should make note within the StorageAccessFramework heading in the docs that the other fs methods can be used with SAF paths, OR, we could alias those methods under the StorageAccessFramework namespace where it makes sense.

<hr />

this is a lot of stuff, let me know if you want to talk through it tmrw

lukmccall committed:
Add `ScopedStorageFramework` namespace

lukmccall commented:
@brentvatne, I've added a `StorageAccessFramework` namespace and functions which create a new file and directory. The main reason why I've added those functions was to write an better example in the NCL.

esamelson commented on docs/pages/versions/unversioned/sdk/filesystem.md:
```diff
@@ -595,24 +595,146 @@ FileSystem.getTotalDiskCapacityAsync().then(totalDiskCapacity => {
+
+## SAF URI
+
+A SAF URI is a URI that is compatible with the Storage Access Framework. It should look like this `content://com.android.externalstorage.*`. The easiest way to obtain such URI is by `askForDirectoryPermissionsAsync` method.
```
```suggestion
A SAF URI is a URI that is compatible with the Storage Access Framework. It should look like this `content://com.android.externalstorage.*`. The easiest way to obtain such URI is by `requestDirectoryPermissionsAsync` method.
```

esamelson commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java:
```diff
@@ -850,6 +937,158 @@ public void downloadResumablePauseAsync(final String uuid, final Promise promise
+        }
+        promise.resolve(result);
+      } else {
+        throw new IOException("Unsupported scheme for location '" + uri + "'.");
```
```suggestion
        throw new IOException("The URI '" + uri + "' is not a Storage Access Framework URI. Try using FileSystem.readDirectoryAsync instead.");
```
etc. for the rest of these functions

esamelson commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java:
```diff
@@ -513,6 +597,9 @@ public void readDirectoryAsync(String uriStr, Map<String, Object> options, Promi
             "Directory '" + uri + "' could not be read.");
         }
+      } else if (isSAFUri(uri)) {
+        promise.reject("ERR_FILESYSTEM_UNSUPPORTED_SCHEME",
+          "Can't read Storage Access Framework directory, use StorageAccessFramework.readDirectoryAsync() instead.");
```
👍

esamelson commented on docs/pages/versions/unversioned/sdk/filesystem.md:
```diff
@@ -595,24 +595,146 @@ FileSystem.getTotalDiskCapacityAsync().then(totalDiskCapacity => {
+
+### `StorageAccessFramework.getUriForDirectoryInRoot(folderName)`
+
+Gets a [SAF URI](#saf-uri) pointing to a folder in the Android root directory. You can use this function to get URI for `StorageAccessFramework.requestForDirectoryPermissionsAsync` when you trying to migrate an album. In that case, the name of the album is the folder name.
```
```suggestion
Gets a [SAF URI](#saf-uri) pointing to a folder in the Android root directory. You can use this function to get URI for `StorageAccessFramework.requestDirectoryPermissionsAsync` when you trying to migrate an album. In that case, the name of the album is the folder name.
```

esamelson replied:
Do you think it's worth showing an example of this usage, either here or in the MediaLibrary docs?

esamelson reviewed (commented):
a few small changes, but it otherwise lgtm 👍

lukmccall committed:
Add example

lukmccall committed:
Add changelog

brentvatne reviewed (approved):
🎉

brentvatne commented:
tests are failing here but not locally 🤔 i rebased master and pushed again to see if that helps

brentvatne committed:
Fix lint error and typo
